source/fastq/fastq.d

```diff
@@ -0,0 +1,322 @@
+ * 15343 	'x'-coordinate of the cluster within the tile
+ * 197393 	'y'-coordinate of the cluster within the tile
+ * 1 	the member of a pair, 1 or 2 (paired-end or mate-pair reads only)
+ * Y 	Y if the read is filtered, N otherwise
+ * 18 	0 when none of the control bits are on, otherwise it is an even number
+ * ATCACG 	index sequence. It is sample number if it is a number
+ *
+ */
+
+module fastaq.fastq;
+
+import iopipe.traits;
+import iopipe.textpipe;
+private import std.traits;
+private import std.range.primitives;
+private import std.algorithm : find, splitter, filter;
+private import std.conv: to;
+private import std.ascii: isWhite;
+private import std.string : stripLeft, stripRight, strip;
+private import std.experimental.logger;
+import fastaq.common.utils;
+
+struct FastqToken
+{
+  size_t endPos;
+  BufRef entryid;
+  BufRef[] hfields;
+  BufRef sequence;
+  BufRef[] extras;
+  BufRef optional;
+  BufRef quality;
+
+  void release(size_t elements)
+  {
+    endPos -= elements;
+    entryid.release(elements);
+    sequence.release(elements);
+    optional.release(elements);
+    quality.release(elements);
+    foreach(ref f; hfields) f.release(elements);
+    foreach(ref f; extras) f.release(elements);
+  }
+
+  auto value(B)(B buf)
+  {
+    FastqConcreteToken!B result;
+    result.entryid = entryid.value(buf);
+    result.hfields = new B[hfields.length];
+    result.extras = new B[extras.length];
+    foreach(i, ref f; hfields)
+      result.hfields[i] = f.value(buf);
+    foreach(j, ref e; extras)
+      result.extras[j] = e.value(buf);
+    result.sequence = sequence.value(buf);
+    result.optional = optional.value(buf);
+    result.quality = quality.value(buf);
+    return result;
+  }
+}
+
+struct FastqConcreteToken(R)
+{
+  R entryid;
+  R[] hfields;
+  R[] extras;
+  R sequence;
+  R optional;
+  R quality;
+
+}
+
+auto tokenParser(Chain,
+                 char fieldsep = ' ',
+                 char subfs = ':')(Chain c) if (isIopipe!Chain && isSomeChar!(ElementEncodingType!(WindowType!Chain)))
+  {
+    auto lines = c.byLine;
+    alias ChainType = typeof(lines);
+    size_t lineCount = 0;
+    static struct Result
+    {
+      ChainType chain;
+      size_t pos;
+      alias chain this;
+
+      FastqToken nextToken()
+      {
+        // pos must start with a start identifier
+        if(pos == chain.window.length)
+          return FastqToken.init;
+        logf("Pos: %s, window length: %s, window: %s", pos, chain.window.length, chain.window );
+        assert(chain.window[pos] == '@', "Got this: " ~ chain.window[pos]);
+        // the header is the current line
+        pos += 1; // jump over the '@'
+        FastqToken result;
+        // parse header line
+        auto lazyArr = chain.window[pos .. $].stripRight.splitter(fieldsep);
+        if(!lazyArr.empty)
+          {
+            auto firstElem = lazyArr.front;
+            result.entryid = BufRef(pos, firstElem.length);
```

Comment:
So looking at this and the format specification, you are first splitting on space, and then sub-splitting on colon. However, you are assuming the first item doesn't have any sub-items. I'm not sure what the intention is, or what the actual file format you are encountering, but this seems to be not correct with the given example above

For instance:

```
@EAS139:136:FC706VJ:2:2104:15343:197393 1:Y:18:ATCACG
```
`entryid` will be set to `@EAS139:136:FC706VJ:2:2104:15343:197393`
and the sub-fields will be parsed as `1`, `Y`, `18`, `ATCACG`.

Not sure if this is what you are looking for, or maybe the actual format looks different in practice?

EDIT: I see that you are indeed parsing the first field as well, but `entryid` will still be set to what I said above, it's just that you will get all the subfields as `hfields` and the other fields as `extras`. I think I see what you are doing here, I think you need to set the `entryid` later, if you are trying to store it as `EAS139`, but maybe I'm wrong.

Reply:
`EAS139` is instrument ID, and it is repeated across all entries and files produced by the instrument. So it is not a good sequence ID. You are right, the entry ID is still `EAS139:136:FC706VJ:2:2104:15343:197393`, and it is correct.  For library users, `hfields` and `extras` are something useful, but not always. In some other case, users may just take the whole header line and focus more onto the `sequence` and `quality`.  So, I am thinking for performance optimization later, we can make header parsing part optional.